isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.prepare.CalciteCatalogReader;
+import org.apache.calcite.prepare.RelOptTableImpl;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexNode;
+
+public class SubstraitRelNodeConverter implements RelVisitor<RelNode, RuntimeException> {
+
+  private final RelOptCluster relOptCluster;
+  private final CalciteCatalogReader catalogReader;
+
+  private final SimpleExtension.ExtensionCollection extensions;
+
+  private final ScalarFunctionConverter scalarFunctionConverter;
+
+  public SubstraitRelNodeConverter(
+      SimpleExtension.ExtensionCollection extensions,
+      RelOptCluster relOptCluster,
+      CalciteCatalogReader catalogReader) {
+    this.relOptCluster = relOptCluster;
+    this.catalogReader = catalogReader;
+    this.extensions = extensions;
+    this.scalarFunctionConverter =
+        new ScalarFunctionConverter(
+            this.extensions.scalarFunctions(), relOptCluster.getTypeFactory());
+  }
+
+  public static RelNode convert(
+      Rel relRoot, RelOptCluster relOptCluster, CalciteCatalogReader calciteCatalogReader) {
+    return relRoot.accept(
+        new SubstraitRelNodeConverter(EXTENSION_COLLECTION, relOptCluster, calciteCatalogReader));
+  }
+
+  @Override
+  public RelNode visit(Aggregate aggregate) throws RuntimeException {
+    throw new UnsupportedOperationException("Aggregate is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(EmptyScan emptyScan) throws RuntimeException {
+    throw new UnsupportedOperationException("EmptyScan is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(Fetch fetch) throws RuntimeException {
+    throw new UnsupportedOperationException("Fetch is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(Filter filter) throws RuntimeException {
+    RelNode input = filter.getInput().accept(this);
+    RexNode filterCondition =
+        filter
+            .getCondition()
+            .accept(
+                new ExpressionRexConverter(
+                    relOptCluster.getTypeFactory(), scalarFunctionConverter));
+    return LogicalFilter.create(input, filterCondition);
+  }
+
+  @Override
+  public RelNode visit(Join join) throws RuntimeException {
+    throw new UnsupportedOperationException("Join is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(NamedScan namedScan) throws RuntimeException {
+    RelDataType rowType =
+        TypeConverter.convert(
+            relOptCluster.getTypeFactory(),
+            namedScan.getRecordType(),
+            namedScan.getInitialSchema().names());
```

Comment:
After switched to using RelBuilder,  looks like I do not have to build row type, as it would be retrieved from relOptSchema when RelBuilder creates a `LogicalScan`. So, the above code is removed.